core/src/main/java/org/apache/iceberg/avro/AssignIds.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.avro;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Map;
+import org.apache.avro.Schema;
+
+/**
+ * Assign field ids to the Avro schema using the given field name to field id map
+ */
+public class AssignIds extends AvroSchemaVisitor<Schema> {
+  // Mapping constructed from table properties
+  private final Map<String, Integer> nameToId;
+  // To be used for maps and list. We do not store ids for maps and lists in our mapping, and it is not
+  // really required. For maps and list we assign a default id.
+  private static final int DEFAULT_ID = -1;
+
+  AssignIds(Map<String, Integer> nameToId) {
+    this.nameToId = nameToId;
+  }
+
+  @Override
+  public Schema record(Schema record, List<String> names, List<Schema> types) {
+    List<Schema.Field> fields = record.getFields();
+    int length = fields.size();
+    List<Schema.Field> newFields = Lists.newArrayListWithExpectedSize(length);
+    for (int i = 0; i < length; i += 1) {
+      newFields.add(newField(fields.get(i), types.get(i)));
+    }
+    return AvroSchemaUtil.copyRecord(record, newFields, null);
+  }
+
+  @Override
+  public Schema map(Schema map, Schema value) {
+    Schema newMap = Schema.createMap(value);
+    newMap.addProp(AvroSchemaUtil.KEY_ID_PROP, DEFAULT_ID);
+    newMap.addProp(AvroSchemaUtil.VALUE_ID_PROP, DEFAULT_ID);
+    return newMap;
+  }
+
+  @Override
+  public Schema array(Schema array, Schema element) {
+    Schema newArray = Schema.createArray(element);
+    newArray.addProp(AvroSchemaUtil.ELEMENT_ID_PROP, DEFAULT_ID);
+    return newArray;
+  }
+
+  @Override
+  public Schema union(Schema union, List<Schema> options) {
+    return Schema.createUnion(options);
+  }
+
+  @Override
+  public Schema primitive(Schema primitive) {
+    return primitive;
+  }
+
+  private Schema.Field newField(Schema.Field field, Schema type) {
+    Schema.Field newField = AvroSchemaUtil.copyField(field, type, field.name());
+    // TODO: Do we need to consider field.aliases()
```

Comment:
Good question. I'm thinking no.

I think this should be maintained by the mapping. Mappings can have more than one name for the same field ID, and that's a better way to handle aliases than trying to keep track of all the possible combinations of aliases in the visitor and testing each one.

This assumes that when an alias is added to the Avro schema, the mapping is also updated. I think we should make that easy so that this doesn't hit users.